### PR TITLE
Update gid API documentation.

### PR DESCRIPTION
**Pull request by hidmic**
Namely, rmw_get_gid_for_publisher() and rmw_compare_gids_equal().

### CODE REVIEW

**Review by Lobotuerk — APPROVED** [review]

File: rmw/include/rmw/rmw.h [hunk 1]
```diff
@@ -2081,23 +2081,69 @@ rmw_count_subscribers(
...
+ *   - Access to primitive data-type arguments is not synchronized.
+ *     It is not safe to read or write `result` while rmw_compare_gids_equal() uses it.
+ *
+ * \param[in] gid1 First unique identifier to compare.
+ * \param[in] gid2 Second unique identifier to compare.
+ * \param[out] bool true if both gids are equal, false otherwise.
+ * \return `RMW_RET_OK` if successful,
+ * \return `RMW_RET_INVALID_ARGUMENT` if `gid1` or `gid2` is NULL, or
+ * \return `RMW_RET_INCORRECT_RMW_IMPLEMENTATION` if `gid1`'s' or `gid2`'s' implementation
+ *   identifier does the match this implementation, or
```
Comment (clalancette) [inline rmw/include/rmw/rmw.h:2145]:
```suggestion
 * \return `RMW_RET_INCORRECT_RMW_IMPLEMENTATION` if the implementation identifier of
 *    `gid1` or `gid2` does not match this implementation, or
```

File: rmw/include/rmw/rmw.h [hunk 2]
```diff
@@ -2081,23 +2081,69 @@ rmw_count_subscribers(
...
+ *   identifier does not match this implementation, or
+ * \return `RMW_RET_ERROR` if an unspecified error occurs.
  */
 RMW_PUBLIC
 RMW_WARN_UNUSED
 rmw_ret_t
 rmw_get_gid_for_publisher(const rmw_publisher_t * publisher, rmw_gid_t * gid);
 
-/// Check if two gid objects are the same
+/// Check if two unique identifiers or gids are equal.
```
Comment (clalancette) [inline rmw/include/rmw/rmw.h:2120]:
```suggestion
/// Check if two unique identifiers (gids) are equal.
```

**Review by clalancette — CHANGES_REQUESTED** [review]
Small nits to fix, nothing major.

File: rmw/include/rmw/rmw.h [hunk 3]
```diff
@@ -2081,23 +2081,69 @@ rmw_count_subscribers(
   const char * topic_name,
   size_t * count);
 
-/// Get the unique identifier of the publisher
+/// Get the unique identifier or gid of a publisher.
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rmw/include/rmw/rmw.h]: Yeap, yeap. See https://github.com/ros2/rmw/pull/274/commits/88995581fe105e753c8df21341ed056ac2d674d2.

File: rmw/include/rmw/rmw.h [hunk 4]
```diff
@@ -2081,23 +2081,69 @@ rmw_count_subscribers(
...
+ * Uses Atomics       | Maybe [1]
+ * Lock-Free          | Maybe [1]
+ *
+ * <i>[1] implementation defined, check implementation documentation.</i>
+ *
+ * \par Thread-safety
+ *   Publishers are thread-safe objects, and so are all operations on them except for
+ *   finalization.
+ *   Therefore, it is safe to get the unique identifier from the same publisher concurrently.
+ *   However, Access to the gid is not synchronized.
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rmw/include/rmw/rmw.h]: See https://github.com/ros2/rmw/pull/274/commits/88995581fe105e753c8df21341ed056ac2d674d2.

File: rmw/include/rmw/rmw.h [hunk 5]
```diff
@@ -2081,23 +2081,69 @@ rmw_count_subscribers(
...
+ *   It is not safe to read or write `gid` while rmw_get_gid_for_publisher() uses it.
+ *
+ * \pre Given `publisher` must be a valid subscription, as returned by rmw_create_publisher().
+ *
+ * \param[in] publisher Publisher to be get a gid from.
+ * \param[out] gid Publisher's unique identifier, populated on success
+ *   but left unchanged on failure.
+ * \return `RMW_RET_OK` if successful, o
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rmw/include/rmw/rmw.h]: Oof, I clearly did this in rush. See https://github.com/ros2/rmw/pull/274/commits/88995581fe105e753c8df21341ed056ac2d674d2.

File: rmw/include/rmw/rmw.h [hunk 6]
```diff
@@ -2081,23 +2081,69 @@ rmw_count_subscribers(
...
+ *   Unique identifier comparison is a reentrant function, but:
+ *   - Access to both gids is read-only but it is not synchronized.
+ *     Concurrent `gid1` and `gid2` reads are safe, but concurrent reads and writes are not.
+ *   - Access to primitive data-type arguments is not synchronized.
+ *     It is not safe to read or write `result` while rmw_compare_gids_equal() uses it.
+ *
+ * \param[in] gid1 First unique identifier to compare.
+ * \param[in] gid2 Second unique identifier to compare.
+ * \param[out] bool true if both gids are equal, false otherwise.
+ * \return `RMW_RET_OK` if successful,
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rmw/include/rmw/rmw.h]: See https://github.com/ros2/rmw/pull/274/commits/88995581fe105e753c8df21341ed056ac2d674d2.

File: rmw/include/rmw/rmw.h (hunk 1 above)
Comment (hidmic) [reply rmw/include/rmw/rmw.h]: See https://github.com/ros2/rmw/pull/274/commits/88995581fe105e753c8df21341ed056ac2d674d2.

File: rmw/include/rmw/rmw.h (hunk 2 above)
Comment (hidmic) [reply rmw/include/rmw/rmw.h]: See https://github.com/ros2/rmw/pull/274/commits/88995581fe105e753c8df21341ed056ac2d674d2.

File: rmw/include/rmw/rmw.h [hunk 7]
```diff
@@ -2081,23 +2081,69 @@ rmw_count_subscribers(
...
+ * \par Thread-safety
+ *   Publishers are thread-safe objects, and so are all operations on them except for
+ *   finalization.
+ *   Therefore, it is safe to get the unique identifier from the same publisher concurrently.
+ *   However, access to the gid is not synchronized.
+ *   It is not safe to read or write `gid` while rmw_get_gid_for_publisher() uses it.
+ *
+ * \pre Given `publisher` must be a valid subscription, as returned by rmw_create_publisher().
+ *
+ * \param[in] publisher Publisher to be get a gid from.
```
Comment (clalancette) [inline rmw/include/rmw/rmw.h:2105]:
```suggestion
 * \param[in] publisher Publisher to get a gid from.
```

**Review by clalancette — APPROVED** [review]
One more nit, approving anyway.

File: rmw/include/rmw/rmw.h (hunk 7 above)
Comment (hidmic) [reply rmw/include/rmw/rmw.h]: Done in https://github.com/ros2/rmw/pull/274/commits/a680b34da1b5be45d946327ffb3f19386e504af1.

**Pull request merged**